From dd939e1e4ad2958f9a4aa2114d09b35b4b8e4a88 Mon Sep 17 00:00:00 2001
From: Leo Yu-Chi Liang <ycliang@andestech.com>
Date: Mon, 12 Aug 2024 20:03:20 +0800
Subject: [PATCH 115/168] smu: andes: pm: Make sure hart0 is the main hart when
 entering sleep mode (#164)

Hart0 and L2 shares the same power domain on Andes 25 series CPU,
so hart0 needs to be the last hart to enter suspend.

We make sure hart0 to be the last hart by making it to call SUSP
sbi call, rather than calling HSM sbi call like other harts do.

Signed-off-by: Leo Yu-Chi Liang <ycliang@andestech.com>
---
 include/linux/cpu.h | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/include/linux/cpu.h b/include/linux/cpu.h
index a7d91a167a8b..6963007d8753 100644
--- a/include/linux/cpu.h
+++ b/include/linux/cpu.h
@@ -170,6 +170,21 @@ static inline int suspend_disable_secondary_cpus(void)
 	if (IS_ENABLED(CONFIG_PM_SLEEP_SMP_NONZERO_CPU))
 		cpu = -1;
 
+	/*
+	 * The power domain of L2 and hart 0 is the same,
+	 * we want to make sure "primary cpu" is mhartid 0,
+	 * so that hart0 would be the last hart to enter
+	 * OpenSBI to disable L2 before entering WFI.
+	 * Note: Kernel cpu_id is not mhartid.
+	 */
+
+	if (IS_ENABLED(CONFIG_ANDES_ATCSMU)) {
+	        for_each_possible_cpu(cpu) {
+	                if (cpuid_to_hartid_map(cpu) == 0)
+	                        break;
+	        }
+	}
+
 	return freeze_secondary_cpus(cpu);
 }
 static inline void suspend_enable_secondary_cpus(void)
-- 
2.34.1

